include/RAJA/policy/openmp/reduce.hpp

```diff
@@ -60,1008 +59,102 @@
-      : m_parent(other.m_parent ? other.m_parent : &other), m_val(other.m_val)
-  {
-  }
-
-  //! Destructor folds value into parent object.
-  RAJA_HOST_DEVICE ~ReduceMax()
-  {
-    if (m_parent) {
-#pragma omp critical
-      {
-        Reduce()(m_parent->m_val, m_val);
-      }
-    }
-  }
-
-  //! return the reduced min value.
-  /*!
-   *  \return the calculated reduced value
-   */
-  RAJA_HOST_DEVICE operator T() {
-    return m_val;
-  }
-
-  //! return the reduced min value.
-  /*!
-   *  \return the calculated reduced value
-   */
-  RAJA_HOST_DEVICE T get() {
-    return operator T();
-  }
-
-  //! reducer function; updates the current instance's state
-  /*!
-   * Assumes each thread has its own copy of the object.
-   */
-  RAJA_HOST_DEVICE const ReduceMax &max(T rhs) const
-  {
-    Reduce()(m_val, rhs);
-    return *this;
-  }
-
-  //! reducer function; updates the current instance's state
-  /*!
-   * Assumes each thread has its own copy of the object.
-   */
-  RAJA_HOST_DEVICE ReduceMax &max(T rhs)
-  {
-    Reduce()(m_val, rhs);
-    return *this;
-  }
-
-private:
-  //! pointer to the parent ReduceMax object
-  const ReduceMax *m_parent;
-  mutable T m_val;
-};
-
-/*!
- **************************************************************************
- *
- * \brief  Sum reducer class template for use in OpenMP execution.
- *
- **************************************************************************
- */
-template <typename T>
-class ReduceSum<omp_reduce, T>
+template <typename T, typename Reduce>
+class ReduceOMP
+    : public reduce::detail::BaseCombinable<T, Reduce, ReduceOMP<T, Reduce>>
 {
-  using Reduce = RAJA::reduce::sum<T>;
+  using Base = reduce::detail::BaseCombinable<T, Reduce, ReduceOMP>;
 
 public:
   //! prohibit compiler-generated default ctor
-  ReduceSum() = delete;
-
-  //! prohibit compiler-generated copy assignment
-  ReduceSum &operator=(const ReduceSum &) = delete;
+  ReduceOMP() = delete;
 
-  //! compiler-generated move constructor
-  ReduceSum(ReduceSum &&) = default;
+  using Base::Base;
 
-  //! compiler-generated move assignment
-  ReduceSum &operator=(ReduceSum &&) = default;
-
-  //! constructor requires a default value for the reducer
-  RAJA_HOST_DEVICE explicit ReduceSum(T init_val, T initializer = T())
-      : m_parent(nullptr), m_val(init_val), m_custom_init(initializer)
+  ~ReduceOMP()
   {
-  }
-
-  //! create a copy of the reducer
-  /*!
-   * keep parent the same if non-null or set to current
-   */
-  RAJA_HOST_DEVICE ReduceSum(const ReduceSum &other)
-      : m_parent(other.m_parent ? other.m_parent : &other),
-        m_val(other.m_custom_init),
-        m_custom_init(other.m_custom_init)
-  {
-  }
-
-  //! Destructor folds value into parent object.
-  RAJA_HOST_DEVICE ~ReduceSum()
-  {
-    if (m_parent) {
-#pragma omp critical
-      {
-        Reduce()(m_parent->m_val, m_val);
-      }
+    if (Base::parent) {
+#pragma omp critical(ompReduceCritical)
```

Comment:
Eventually will we want to use locks? Or is the named critical section enough still improve our current performance?

Reply:
I actually tried to use locks, it worked great with everything but GOMP, which would segfault whenever I tried to `new` an omp_lock_t or any structure containing an omp_lock_t.  It would be great to track that down and switch over, but seemed like something for a subsequent PR.